mllib/src/main/scala/org/apache/spark/ml/evaluation/RankingEvaluator.scala

```diff
@@ -0,0 +1,287 @@
+ */
+@Since("2.0.0")
+@Experimental
+final class RankingEvaluator[T: ClassTag] @Since("2.0.0") (@Since("2.0.0") override val uid: String)
+  extends Evaluator with HasPredictionCol with HasLabelCol with DefaultParamsWritable with Logging {
+
+  @Since("2.0.0")
+  def this() = this(Identifiable.randomUID("rankingEval"))
+
+  @Since("2.0.0")
+  final val k = new IntParam(this, "k", "Top-K cutoff", (x: Int) => x > 0)
+
+  /** @group getParam */
+  @Since("2.0.0")
+  def getK: Int = $(k)
+
+  /** @group setParam */
+  @Since("2.0.0")
+  def setK(value: Int): this.type = set(k, value)
+
+  setDefault(k -> 1)
+
+  /**
+   * Param for metric name in evaluation. Supports:
+   *  - `"map"` (default): Mean Average Precision
+   *  - `"mapk"`: Mean Average Precision@K
+   *  - `"ndcg"`: Normalized Discounted Cumulative Gain
+   *  - `"mrr"`: Mean Reciprocal Rank
+   *
+   * @group param
+   */
+  @Since("2.0.0")
+  val metricName: Param[String] = {
+    val allowedParams = ParamValidators.inArray(Array("map", "mapk", "ndcg", "mrr"))
+    new Param(this, "metricName", "metric name in evaluation (map|mapk|ndcg||mrr)", allowedParams)
+  }
+
+  /** @group getParam */
+  @Since("2.0.0")
+  def getMetricName: String = $(metricName)
+
+  /** @group setParam */
+  @Since("2.0.0")
+  def setMetricName(value: String): this.type = set(metricName, value)
+
+  /** @group setParam */
+  @Since("2.0.0")
+  def setPredictionCol(value: String): this.type = set(predictionCol, value)
+
+  /** @group setParam */
+  @Since("2.0.0")
+  def setLabelCol(value: String): this.type = set(labelCol, value)
+
+  setDefault(metricName -> "map")
+
+  @Since("2.0.0")
+  override def evaluate(dataset: Dataset[_]): Double = {
+    val schema = dataset.schema
+    val predictionColName = $(predictionCol)
+    val predictionType = schema($(predictionCol)).dataType
+    val labelColName = $(labelCol)
+    val labelType = schema($(labelCol)).dataType
+    require(predictionType == labelType,
+      s"Prediction column $predictionColName and Label column $labelColName " +
+        s"must be of the same type, but Prediction column $predictionColName is $predictionType " +
+        s"and Label column $labelColName is $labelType")
+
+    val metric = $(metricName) match {
+      case "map" => meanAveragePrecision(dataset)
+      case "ndcg" => normalizedDiscountedCumulativeGain(dataset)
+      case "mapk" => meanAveragePrecisionAtK(dataset)
+      case "mrr" => meanReciprocalRank(dataset)
+    }
+    metric
+  }
+
+  /**
+   * Returns the mean average precision (MAP) of all the queries.
+   * If a query has an empty ground truth set, the average precision will be zero and a log
+   * warning is generated.
+   */
+  private def meanAveragePrecision(dataset: Dataset[_]): Double = {
+    val sc = SparkContext.getOrCreate()
+    val sqlContext = SQLContext.getOrCreate(sc)
+    import sqlContext.implicits._
+
+    dataset.map{ case (prediction: Array[T], label: Array[T]) =>
```

Comment:
This may easily fail if the input dataset doesn't only have these columns. Typically above we would do a `select($(predictionCol), $(labelCol))` call.